Hotfix/fix missing client capabilities

morantr opened the pull request:
@amritak23 has discovered issues on weighted phyrate while testing the platform.
The bug root cause was on the parsing of the client capabilities which occurred the function that generates connect events to a client that are connected to the ap before beerocks has brought up.
Also, there was a bug on the bandwidth enum on the station capabilities report.

morantr committed:
controller/bwl: Add prints

Add useful prints that were missing such as vht_mcs on bwl, or others that
can make debugging much easier such as in the oprimal_path_task and db.cpp.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
bwl: Change stoi convertor on dwpal

The beeocks stoi function does not support converting numbers on a non-decimal base.
Therefore, the conversion of VHT/HT MCS was always 0.
This led to getting the wrong value when looking at the MCS Table.

Change the stoi converter to the one on std namespace which is used wildly on this file.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
bwl: Align eWiFiBandwidth to beerocks_defines

On commit d90f2403, eWiFiBandwidth on beerocks_defines was changed,
but the parallel definition on bwl did not.
This led to a wrong bandwidth on the station capabilities report, which affected
the optimal_path result.

Align the eWiFiBandwidth enum definition on bwl to its parallel on beerocks_defines.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Always update station capabilities

Change the station capabilities setter to always update the station capabilities,
and not just once.
I found out that the first capabilities that were saved on the database were sourced
on the function that generates client connect events and not by a real sta-connect event.
The capabilities parsing on that function is a little bit different, and potentially, if there is
a bug in the capabilities parsing like what has been fixed on one of the previous commits,
the capabilities will be wrong forever on the database.

Also, if a client updates its driver, its capabilities could change so it is always a good idea
to update them.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

adam1985d commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -420,7 +420,9 @@ void optimal_path_task::work()
             if (!force_signal_strength_decision &&
                 !database.settings_client_optimal_path_roaming_prefer_signal_strength()) {
-                //get sta capabilities....
+                // Get sta capabilities
```
nitpick:
no need for this comment once you added the log print

adam1985d commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -922,7 +924,9 @@ void optimal_path_task::work()
             }
 
-            //get sta capabilities....
+            // Get STA capabilities
```
nitpick:
no need for this comment once you added the log print

adam1985d commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1064,26 +1064,14 @@ bool db::set_station_capabilities(const std::string &client_mac,
-        }
-
-        n->capabilities = n->m_sta_5ghz_capabilities;
+        n->m_sta_5ghz_capabilities          = sta_cap;
```
is there a reason the original implementation didn't update the capabilities?
is there a side-effect to the set that was meant to be prevented

morantr replied:
> is there a reason the original implementation didn't update the capabilities?

Yes, to prevent capabilities struct copy.

adam1985d commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1064,26 +1064,14 @@ bool db::set_station_capabilities(const std::string &client_mac,
-        }
-
-        n->capabilities = n->m_sta_24ghz_capabilities;
+        n->m_sta_24ghz_capabilities          = sta_cap;
```
same question here

kantera800 commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1064,26 +1064,14 @@ bool db::set_station_capabilities(const std::string &client_mac,
-
-        n->capabilities = n->m_sta_5ghz_capabilities;
+        n->m_sta_5ghz_capabilities          = sta_cap;
+        n->m_sta_5ghz_capabilities.cap_flag = true;
```
as this commit headline  implies you can go ahead and remove cap_flag

morantr replied:
Actually no. When this flag is not set, we return nullptr on capability getter function.

vitalybu replied:
So it determined the validity of the data?
In that case, maybe should be changed to `valid`

morantr replied:
rename it on a separate commit.

vitalybu commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -372,14 +372,14 @@ static std::shared_ptr<char> generate_client_assoc_event(const std::string &even
     for (uint i = 0; (i < 8) && (i < sizeof(ht_mcs)); i++) {
         ht_mcs_str[2] = ht_mcs[2 * i];
         ht_mcs_str[3] = ht_mcs[2 * i + 1];
-        HT_MCS[i]     = beerocks::string_utils::stoi(ht_mcs_str);
```
There's a reason `string_utils::stoi` was used here and not `std::stoi`.
The `std` versions throw an `invalid_argument` or `out_of_range` exceptions if the input is not valid.
Since we don't catch exceptions, it'll simply crash the process...
If you really need `stoul`, please implement a safe version.

morantr replied:
add a commit that replaces all calls across the code to safe functions.

vitalybu commented on common/beerocks/bwl/include/bwl/base_wlan_hal_types.h:
```diff
@@ -127,9 +127,11 @@ typedef struct {
 //ap_wlan_hal
 
 enum eWiFiBandwidth : uint8_t {
```
Why don't we use the `eWiFiBandwidth` from `beerocks_defines`?

morantr replied:
Good question! And why don't we use the struct from tlvf? They are the same, and we have a duplication for every message.
The reason is that we didn't want to create dependency (or dependency of messages and types, not sure) from other libs such BCL.
This is really stupid IMO, since we already depended on BCL for some utilities.
@tomereli

morantr committed:
Replace string to integer converter across the repository

On our repository, we are not using try/catch. This is mean that we shall not use
functions that throw exceptions.
There are many lines of code that use some of CPP string to integer converters
which might throw an exception on in case an error has occurred, but we don't
catch, so it the process will crash.

To avoid it, replace all the calls to cstdlib converters which are safe.
Yes, these are C function, but it turns out that CPP converters are just wrappers
of the function on cstdlib so it is ok.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
bcl: Rename sRadioCapabilities "cap_flag" to "valid"

The new name reflects better the meaning of the struct member.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
cppcheck: Update existing issues file

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
